### ISSUE TITLE
Add password authentication method

### ISSUE THREAD

**Issue by thomascube** [linagora/jmap-client#8]
Beside the external authentication, the client should also implement a method `authPassword()` to perform a password authentication as described in http://jmap.io/spec.html#getting-an-access-token

A continuation callback might be used to pass the `prompt` parameter to the caller and get the password back.

**Comment by thomascube** [comment]
Let me rephrase that feature request a little bit: I think the intention of JMAPs 2-step authentication mechanism is to let the client learn about the authentication methods a certain server supports (step one) and then react accordingly towards the user and finally complete authentication at the JMAP server (step 2).

Therefore, implementing the various authentication methods in separate functions like `authExternal()` and `authPassword()` doesn't make sense as this requires the client to decide about the executed method even before taking the first step.

Thus, I suggest to instead implement a generic `authenticate()` method which executes the first step. The continuation callback is called with the list of methods offered by the client and can eventually resolve with the method chosen by the application (or the user) and additional parameters to submit (i.e. the password for `method == password`). This finally executes the second step of the authentication procedure and resolves with an `AuthAccess` object once the server confirms the successful authentication.

Makes sense?

**Comment by thomas-hilaire** [comment]
Hello,

Thank to have opened this issue. We agree that a generic `authenticate` method makes sense and the way that you've drafted it.

We also think that shorthand methods like `authExternal` and `authPassword` should be available in this lib (which delegate to the generic `authenticate`). A regular application may support only one of these `methods` so, we want this library handling the potential lack of it by the JMAP server.

As a side note, we want to add some code style rules that can affect your current code, see #9.

Looking toward your pull-request ! 
Cheers,

**Comment by thomascube** [comment]
Agreed. So I'll rewrite `authExternal` to call the new `authenticate` method and add another shortcut method `authPassword` to do the same. Pull request will follow as soon as I also have written tests for these changes.

Feel free to comment on my commits where you see the coding guidelines violated.

**Comment by thomas-hilaire** [comment]
Fixed by https://github.com/linagora/jmap-client/pull/10
